:memo: Rework services overview (#2042)
platformsh/platformsh-docs: docs/src/configuration/services/_index.md

---
title: Add services
weight: 3
description: See how to add services such as databases, cache, and search engines and configure them to suit your needs.
layout: single
---

Platform.sh includes many services, so you don't have to subscribe to external cache or search engine services.
Because the services are included in your project, you can manage them through Git
and they're backed up together with the rest of your project.

Your project requires a service configuration file named `services.yaml` in the `.platform` directory.
If you don't need any services (such as for a static website), you can leave the file blank.
Read on to see how to add services.

![Services](/images/management-console/relationships.png "0.50")

## Add a service

> Adding a service is a two-step process.

Reviewer: This 2 step overview is priceless, thanks so much @CollierCZ

## 1. Configure the service

> All service configuation happens in the configuration file (`.platform/services.yaml`) in your Git repository.

Reviewer: ```suggestion
All service configuration happens in the configuration file (`.platform/services.yaml`) in your Git repository.
```

Configure your service in the following pattern:

```yaml
<SERVICE_NAME>:
    type: <SERVICE_TYPE>:<VERSION>
    # Other options...
```

An example service configuration for two databases might look like this:

```yaml
database1:
    type: mariadb:10.5
    disk: 2048
database2:
    type: postgresql:13
    disk: 1024
```

This YAML file is a dictionary defining all of the services you want to use.

> The top-level key is the service name, which you use to identify the service in step 2.

Reviewer: ```suggestion
The top-level key is a custom service name, which you use to identify the service in step 2.
```

You can give it any name you want with lowercase alphanumeric characters, hyphens, and underscores.

{{< note >}}

Changing the service name is interpreted as creating an entirely new service.
This **removes all data in that service**.

> Always back up your data before changing the file.

Reviewer: ```suggestion
Always back up your data before changing existing services in your `.platform/services.yaml` file.
```

### Service options

The following table presents the keys you can define for each service:

| Name            | Type        | Required          | Description |
| --------------- | ----------- | ----------------- | ----------- |
| `type`          | `string`    | Yes               | One of the [available services](#available-services) in the format `type:version`. |
| `disk`          | `integer`   | For some services | The size in MB of the [persistent disk](#disk) allocated to the service. Can't be set for memory-resident-only services such as `memcache` and `redis`. |
| `size`          | `string`    |                   | How many CPU and memory [resources to allocate](#size) to the service. Possible values are `AUTO` (default), `S`, `M`, `L`, `XL`, `2XL`, and `4XL`. |
| `configuration` | dictionary  |                   | Some services have additional specific configuration options that can be defined here, such as specific endpoints. See the given service page for more details. |

Both `disk` and `size` are limited by your plan settings.

#### Disk

Downsizing a service's persistent disk isn't currently supported
in the `eu.platform.sh` and `us.platform.sh` regions.

> If necessary, [migrate your project](../../guides/general/region-migration.md).

Reviewer: I'm so tempted to include a note to migrate either way, but it's okay I guess 😏

#### Size

By default, Platform.sh allocates CPU and memory resources to each container automatically given the available resources on the plan.
Some services are optimized for high CPU load, some for high memory load.

> If your plan is sufficiently large, you can customize the size of your service container.

Reviewer: ```suggestion
If your plan is sufficiently large and fits bigger containers, you can increase the size of your service container.
```
I'm trying to help, but I'm not sure it does, feel free to reword from here. The main goal is to spell that sufficiently large means that all containers need to fit in the plan 😬  might be okay as is though :)

Author: I saw the original idea as including the possibility of making a given service smaller. Like if your plan isn't super big and you want to give one service more room, you could maybe make service X smaller to make room for service Y.

But then the "sufficiently large" part makes less sense. :thinking:

### 2. Connect the service

Once you have configured a service, you need to create a relationship to connect it to an app.
This is done in your [app configuration](../app/_index.md) under the `relationships` key.

The relationship follows this pattern:

```yaml
relationships:
    <RELATIONSHIP_NAME>: "<SERVICE_NAME>:<ENDPOINT>"
```

An example relationship to connect to the databases given in the [example in step 1](#1-configure-the-service):

```yaml
relationships:
    mysql_database: "database1:mysql"
    postgresql_database: "database2:postgresql"
```

As with the service name, you can give the relationship any name you want
with lowercase alphanumeric characters, hyphens, and underscores.
It helps if the service name and relationship name are different, but it isn't required.

Each service offers one or more endpoints for connections, depending on the service.
An endpoint is a named set of credentials to give access to other apps and services in your project.
If you don't specify one in the [service configuration](#service-options), a default endpoint is created.

> The default varies by service, generally being its type (such as `mysql` or `solr`).

Reviewer: ```suggestion
The default endpoint varies by service, generally being its type (such as `mysql` or `solr`).
```

Each service offers one or more endpoints for connections, depending on the service.
An endpoint is a named set of credentials to give access to other apps and services in your project.
If you don't specify one in the [service configuration](#service-options), a default endpoint is created.
The default varies by service, generally being its type (such as `mysql` or `solr`).

## Using the services
## Available services

In order for a service to be available to an application in your project (Platform.sh supports not only multiple backends but also multiple applications in each project) you will need to refer to it in the [.platform.app.yaml](/configuration/app/_index.md) file which configures the *relationships* between applications and services.
The following table presents the available service types and their versions.
Add them to the `type` key of the [service configuration](#1-configure-the-service) in the format `type:version`.

## Endpoints
<!-- To update the versions in this table, use docs/data/registry.json -->
{{< readFile file="src/registry/images/tables/services_supported.md" markdownify="true">}}

> All services offer one or more `endpoints`.  An endpoint is simply a named set of credentials that can be used to give access to other applications and services in your project to that service.  Only some services support multiple user-defined endpoints.  If you do not specify one then one will be created with a standard defined name, generally the name of the service type (e.g., `mysql` or `solr`).  An application container, defined by a `.platform.app.yaml` file, always exposes an endpoint named `http` to allow the [router](/configuration/routes/_index.md) to forward requests to it.

Reviewer: Do we need to keep these details? 
>An application container, defined by a `.platform.app.yaml` file, always exposes an endpoint named `http` to allow the [router](/configuration/routes/_index.md) to forward requests to it.

We do show the router in site architecture graphs, and it's more a service container than an app container, from my PoV at least 🤔

Author: I would think the details about the router and the app container endpoint would fall under the routing description. Mostly, I think it's not actionable on the user's end. It's a description of what is happening in the background, but I don't think it affects anything the user does. Or I didn't see where it would. If it does, I'm happy to add the details about the app container endpoint back.

The other details appear later.

Reviewer: You're right, the router container is not something folks can interact with, at least atm. It's though important for them to understand this concept at some point, since they do see it as part of the architecture of their environment. 🤔  I'm comfortable leaving it as is and have this come up when the router is more visible

> Access to the database or other services is only available from within the cluster.

Reviewer: ```suggestion
Access to the database or other services is only available from within the app.
```
I'm not sure what this meant before, sounds like it was meant for dedicated clusters?

Author: My understanding was it was to cover multi-app projects and such. But I agree the wording could be clearer.

> To connect to a service, you need the [service credentials](#obtain-service-credentials).

Reviewer: shall we specify the specific credentials we require here?

Author: I'm not entirely sure what you mean. Do you think the example isn't clear enough? Would you specify it here rather than in the section below?

Reviewer: Right, I get the feeling that this phrase is the intro to the 2 steps below but it's offering them the chance to jump ahead steps or have them be disconnected. 

Maybe just a 1. Obtain service credentials and 2. Connect to an SSH tunnel would address this?

Author: Yes, that's clearly the right approach. Just like above.

Author: Should be clearer now, @MariadeAnton :crossed_fingers:

Reviewer: Definitely is! Thanks so much for bearing with all the little things here and there. Fantastic work all around, @CollierCZ !

        port: 3306
        host_mapped: false
        url: 'mysql://user:@database.internal:3306/main'
```

With this example, you can connect to the `database` relationship
with the user `user`, an empty password, and the database name `main` (from the `path`).

> ### Connect to the tunnel

Reviewer: ```suggestion
### Connect to the SSH tunnel
```

Might this connect the dots between SSH and the tunnel ?

> Once you have [opened a tunnel](../../development/ssh/_index.md#connect-to-services), you can use it to connect to your service.

Reviewer: ```suggestion
Once you have [opened an SSH tunnel](../../development/ssh/_index.md#connect-to-services), you can use it to connect to your service.
```
